# Hola Alejandro!

Mi nombre es David Bautista, soy code reviewer de TripleTen y voy a revisar el proyecto que acabas de desarrollar.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión.

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres.**

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div

¡Empecemos!

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General

~~Hola, Alejandro, te felicito por el buen desarrollo del proyecto hasta el momento. Ahora bien, he dejado un comentario en rojo para que lo puedas tener en cuenta para la siguiente entrega. Quedo atento.~~ </div>

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div> 
    Buenas tardes David, muchas gracias por tus comentarios, ya agregué el .describe() a la revisión del dataframe como fue solicitado. 
    Saludos!

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General #2

Hola, Alejandro, te felicito por la culminación del proyecto. Muy buen trabajo. </div>

# Proyecto 8

## Introducción.

El objetivo del proyecto es desarrollar un modelo que pueda analizar el comportamiento de los clientes y recomendar uno de los nuevos planes de Megaline: Smart o Ultra, para ello se tiene acceso a los datos de comportamiento de los suscriptores (llamadas, minutos, mensajes, mb usados) que ya se han cambiado a los planes nuevos. 

El umbral de exactitud solicitado es de 0.75.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo, Alejandri. Desarrollar una sección de introducción del proyecto siempre es importante para poder generar un contenido con altos estándares de calidad.
</div>

## Importación de Librerias e Importación del Dataframe

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
import numpy as np 



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo importando las librerías necesarias para el desarrollo del proyecto. 
</div>

In [2]:
megaline = pd.read_csv("/datasets/users_behavior.csv")

In [3]:
megaline.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [4]:
print(megaline.info())
print()
print(megaline.duplicated().sum())
print()
print(megaline.isna().sum())
print()
print(megaline.describe())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None

0

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~Buen trabajo realizando la exploración inicial del contenido de la tabla, aun así , podrías agregar elementos adicionales como el uso del método ``describe()``.~~
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Genial, buen trabajo.
</div>

### Conclusiones.

Se importaron las librerias necesarias y se revisó de forma general el dataframe, los tipos de datos son correctos y no se encuentran valores nulos ni filas repetidas. de igual manera se incluyo un .describe() para poder tener una vision general de algunos parámetros estadísticos dentro del dataframe. 

## Segmentación de los datos.


In [5]:
target= "is_ultra"
features= ["calls", "minutes", "messages", "mb_used"]

In [6]:
train, temp = train_test_split(megaline, test_size=0.4, random_state=8)
val, test = train_test_split(temp, test_size=0.4, random_state=8)

In [7]:
print(train.shape)
print(val.shape)
print(test.shape)

(1928, 5)
(771, 5)
(515, 5)


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo realizando el split de los datos en los sub sets necesarios para el desarrollo correcto de los modelos. 
</div>

In [8]:
X_train = train.drop(columns=['is_ultra'])
y_train = train['is_ultra']

X_val = val.drop(columns=['is_ultra'])
y_val = val['is_ultra']

X_test = test.drop(columns=['is_ultra'])
y_test = test['is_ultra']

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo realizando la separación de las características y los targets. 
</div>

### Conclusiones. 
Se identiticó el target en la columna "is_ultra", quedando las demás columnas para realizar el aprendizaje, de igual manera se segmento como es requerido en el ejercicio en 3 conjuntos, uno de entrenamiento (60%, 1928 filas) uno de validación (24%, 711 filas ) y uno de prueba (16%, 515 filas). Esta división se basa en la necesidad de contar con el grupo de entrenamiento como el mayoritario y un grupo de validación ligeramente mayor que el grupo de prueba.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo. 
</div>

## Modelo de aprendizaje. (Random Forest Classifier)

### Random Forest Clasiffier

In [9]:
model = RandomForestClassifier(random_state=8)
model.fit(X_train, y_train)

y_val_pred = model.predict(X_val)

val_accuracy = accuracy_score(y_val, y_val_pred)
print(f'Accuracy de validación: {val_accuracy}')

print()

print(classification_report(y_val, y_val_pred))


Accuracy de validación: 0.8054474708171206

              precision    recall  f1-score   support

           0       0.83      0.91      0.87       541
           1       0.73      0.55      0.63       230

    accuracy                           0.81       771
   macro avg       0.78      0.73      0.75       771
weighted avg       0.80      0.81      0.80       771



### Logistic Regression 

In [10]:
model = LogisticRegression(random_state=8)

model.fit(X_train, y_train)

y_val_pred = model.predict(X_val)

val_accuracy = accuracy_score(y_val, y_val_pred)
print(f'Accuracy de validación: {val_accuracy}')

print()

print(classification_report(y_val, y_val_pred))

Accuracy de validación: 0.7107652399481194

              precision    recall  f1-score   support

           0       0.72      0.97      0.83       541
           1       0.59      0.10      0.16       230

    accuracy                           0.71       771
   macro avg       0.66      0.53      0.49       771
weighted avg       0.68      0.71      0.63       771



### Decision Tree Classifier

In [11]:
model = DecisionTreeClassifier(random_state=8)
model.fit(X_train, y_train)
y_val_pred = model.predict(X_val)

val_accuracy = accuracy_score(y_val, y_val_pred)
print(f'Accuracy de validación: {val_accuracy}')

print()

print(classification_report(y_val, y_val_pred))


Accuracy de validación: 0.7380025940337225

              precision    recall  f1-score   support

           0       0.82      0.80      0.81       541
           1       0.56      0.60      0.58       230

    accuracy                           0.74       771
   macro avg       0.69      0.70      0.69       771
weighted avg       0.74      0.74      0.74       771



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo desplegando y aplicando las métricas de evaluación para cada uno de los modelos.  
</div>

## Hiperparámetros y Calidad del Modelo.

### Random Forest Classifier

In [12]:
param_grid = {
    'n_estimators': [40, 60, 80],
    'max_depth': [2, 5, 10]
}
model = RandomForestClassifier(random_state=8)

grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, n_jobs=-1, scoring='accuracy')

grid_search.fit(X_val, y_val)

print(f'Mejores hiperparámetros: {grid_search.best_params_}')

print()

print(f'Mejor exactitud de validación: {grid_search.best_score_}')

Mejores hiperparámetros: {'max_depth': 5, 'n_estimators': 40}

Mejor exactitud de validación: 0.8028534370946822


In [13]:
best_model = grid_search.best_estimator_

y_val_pred = best_model.predict(X_val)

val_accuracy = accuracy_score(y_val, y_val_pred)

print(f'Accuracy de prueba del mejor modelo: {val_accuracy}')

print()

print(classification_report(y_val, y_val_pred))

Accuracy de prueba del mejor modelo: 0.8612191958495461

              precision    recall  f1-score   support

           0       0.85      0.98      0.91       541
           1       0.91      0.59      0.72       230

    accuracy                           0.86       771
   macro avg       0.88      0.78      0.81       771
weighted avg       0.87      0.86      0.85       771



### Logistic Regresion.

In [14]:
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10],
    'penalty': ['l1', 'l2'],
}

model = LogisticRegression(random_state=8, max_iter=200, solver="liblinear")

grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='accuracy', verbose=1, n_jobs=-1)

grid_search.fit(X_val, y_val)

print(f'Mejores hiperparámetros: {grid_search.best_params_}')

print()

print(f'Mejor exactitud de validación: {grid_search.best_score_}')

best_model = grid_search.best_estimator_
y_val_pred = best_model.predict(X_val)
test_accuracy = accuracy_score(y_val, y_val_pred)

print(f'Accuracy de prueba del mejor modelo: {test_accuracy}')

print()

print(classification_report(y_val, y_val_pred))

Fitting 5 folds for each of 10 candidates, totalling 50 fits
Mejores hiperparámetros: {'C': 1, 'penalty': 'l1'}

Mejor exactitud de validación: 0.7561374109761206
Accuracy de prueba del mejor modelo: 0.7600518806744487

              precision    recall  f1-score   support

           0       0.75      0.98      0.85       541
           1       0.86      0.23      0.37       230

    accuracy                           0.76       771
   macro avg       0.80      0.61      0.61       771
weighted avg       0.78      0.76      0.71       771



### Decision Tree Classifier

In [15]:
param_grid = {
    'max_depth': [3, 5, 7, 10],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
}

model = DecisionTreeClassifier()
# Configurar GridSearchCV
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='accuracy', verbose=1, n_jobs=-1)

# Ajustar GridSearchCV con los datos de entrenamiento
grid_search.fit(X_train, y_train)

# Mostrar los mejores hiperparámetros y la mejor precisión
print(f'Mejores hiperparámetros: {grid_search.best_params_}')
print(f'Mejor exactitud de validación: {grid_search.best_score_}')

# Evaluar el mejor modelo en el conjunto de prueba
best_model = grid_search.best_estimator_
y_val_pred = best_model.predict(X_val)
val_accuracy = accuracy_score(y_val, y_val_pred)
print(f'Accuracy de prueba del mejor modelo: {val_accuracy}')
print(classification_report(y_val, y_val_pred))


Fitting 5 folds for each of 36 candidates, totalling 180 fits
Mejores hiperparámetros: {'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 5}
Mejor exactitud de validación: 0.796156382477626
Accuracy de prueba del mejor modelo: 0.8067444876783398
              precision    recall  f1-score   support

           0       0.81      0.94      0.87       541
           1       0.79      0.48      0.60       230

    accuracy                           0.81       771
   macro avg       0.80      0.71      0.74       771
weighted avg       0.80      0.81      0.79       771



<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Excelente, Alejandro, buen trabajo desarrollando los modelos de manera más robusta por medio del uso de las grillas de optimización de hiperametros para después también tener el reporte de métricas de calidad. </div>

### Conclusiones.

Al buscar los mejores hiperparámetros para los tres modelos, encontramos que Random Forest Clasiffier es el mejor modelo con max_depth=5 y n_estimators=40, estas pruebas se realizaron en el conjunto de validación, el umbral de exactitud de este modelo se ubica en .80 cumpiendo con las características requeridas, siendo necesario mencionar que los tres modelos alcanzaron el mínimo requerido de 0.75

## Prueba del Modelo.

### Random Forest Clasiffier.

In [16]:
model = RandomForestClassifier(random_state=8, max_depth=5, n_estimators=40)

model.fit(X_test, y_test)

# Predecir con el conjunto de prueba (X_test)
y_test_pred = model.predict(X_test)

test_accuracy = accuracy_score(y_test, y_test_pred)
print(f'Accuracy de prueba: {test_accuracy}')

print(classification_report(y_test, y_test_pred))

Accuracy de prueba: 0.8485436893203884
              precision    recall  f1-score   support

           0       0.83      0.99      0.90       360
           1       0.94      0.53      0.68       155

    accuracy                           0.85       515
   macro avg       0.89      0.76      0.79       515
weighted avg       0.86      0.85      0.83       515



### Conclusiones.
Se puede observar un incremento del umbral de exactitud entre el conjunto de validación y el conjunto de prueba, lo que sugiere que el modelo no se encuentra sobreajustado y puede llegar a ser funcional para la predicción solicitada. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Perfecto, buen trabajo con el desarollo de la prueba final del modelo. </div>

## Prueba de Cordura. 

In [17]:
accuracy_score(test[target], pd.Series([0] * len(test)))

0.6990291262135923

In [18]:
accuracy_score(test[target], pd.Series([1] * len(test)))

0.30097087378640774

In [19]:
y_test_pred_random = np.random.choice([0, 1], size=len(test))

test_accuracy_random = accuracy_score(test[target], y_test_pred_random)

print(test_accuracy_random)

0.5048543689320388


### Conclusión.
Se desafió el modelo Random Forest Clasiffier con hiperparámetros de max_depth=5, n_estimators=40 frente a dos escenarios donde todos los resultados son homogeneos (0 y 1) asi como uno con números aleatorios para verificar que el umbral de exactitud sea superior (0.84 vs 0.69, 0.30, 0.51), concluyendo que el modelo es satisfactorio en estas pruebas. 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Excelente, buen trabajo con el desarrollo de las secciones de prueba de cordura y de conclusiones generales. </div>